src/v/kafka/client/client.cc

```diff
@@ -541,21 +541,34 @@ ss::future<kafka::fetch_response> client::consumer_fetch(
                 std::chrono::duration_cast<std::chrono::milliseconds>(timeout),
                 max_bytes);
           })
-          .then([this](kafka::fetch_response res) {
+          .then([](kafka::fetch_response res) {
+              error_code last_ec{error_code::none};
+              model::topic_partition last_tp{
+                model::topic{}, model::partition_id{-1}};
               bool has_error = std::any_of(
                 res.data.topics.begin(),
                 res.data.topics.end(),
-                [](auto const& topics) {
+                [&last_ec, &last_tp](auto const& topics) {
                     return std::any_of(
                       topics.partitions.begin(),
                       topics.partitions.end(),
-                      [](const auto& p) {
-                          return p.error_code != error_code::none;
+                      [&last_ec, &last_tp, name{topics.name}](const auto& p) {
+                          if (p.error_code != error_code::none) {
+                              last_ec = p.error_code;
+                              last_tp = model::topic_partition{
+                                name, p.partition_index};
+                              return true;
+                          } else {
+                              return false;
+                          }
                       });
                 });
-              return (has_error ? _wait_or_start_update_metadata() : ss::now())
-                .then(
-                  [res{std::move(res)}]() mutable { return std::move(res); });
+
+              if (has_error) {
+                  return ss::make_exception_future<fetch_response>(
+                    partition_error(last_tp, last_ec));
```

Comment:
This could be expressed more clearly by replacing `any_of` by a loop and a throw.

It doesn't do that because it means throwing away some of the results that have been fetched already.

The consumer typically sits in a loop fetching, so the previous behaviour is sufficient, and is able to make progress faster.